# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1479914226 18000
#      Wed Nov 23 10:17:06 2016 -0500
# Node ID 86b6f79d4de19998c555dfe4ce839fef51e26efc
# Parent  f55c208a31c18b750ca599aab28a753c21df9e3b
better compatibility for printf with empty args (bug #39375)

* oct-stream.cc (printf_value_cache::get_next_value):
If argument has no elements, return empty matrix.
(octave_base_stream::do_printf): Handle empty matrix in numeric
conversion case.
* io.tst: New tests.

diff --git a/libinterp/corefcn/oct-stream.cc b/libinterp/corefcn/oct-stream.cc
--- a/libinterp/corefcn/oct-stream.cc
+++ b/libinterp/corefcn/oct-stream.cc
@@ -5424,19 +5424,23 @@ printf_value_cache::get_next_value (char
         }
       else
         {
           val_idx++;
           have_data = false;
 
           if (n_elts == 0)
             {
-              if (elt_idx == 0 && (type == 's' || type == 'c'))
+              if (elt_idx == 0)
                 {
-                  retval = "";
+                  if (type == 's' || type == 'c')
+                    retval = "";
+                  else
+                    retval = Matrix ();
+
                   break;
                 }
 
               if (exhausted ())
                 curr_state = conversion_error;
             }
         }
     }
@@ -5795,18 +5799,21 @@ octave_base_stream::do_printf (printf_fo
               else
                 break;
             }
           else
             {
               octave_value val = val_cache.get_next_value ();
 
               if (val_cache)
-                retval += do_numeric_printf_conv (os, elt, nsa, sa_1,
-                                                  sa_2, val, who);
+                {
+                  if (! val.is_empty ())
+                    retval += do_numeric_printf_conv (os, elt, nsa, sa_1,
+                                                      sa_2, val, who);
+                }
               else
                 break;
             }
 
           if (! os)
             {
               error ("%s: write error", who.c_str ());
               break;
diff --git a/test/io.tst b/test/io.tst
--- a/test/io.tst
+++ b/test/io.tst
@@ -722,8 +722,16 @@
 %!assert (sprintf ("|%x|", "Octave"), "|4f||63||74||61||76||65|")
 %!assert (sprintf ("|%X|", "Octave"), "|4F||63||74||61||76||65|")
 %!assert (sprintf ("|%o|", "Octave"), "|117||143||164||141||166||145|")
 
 ## bug #47192
 %!assert (sprintf ("%s", repmat ("blah", 2, 1)), "bbllaahh")
 %!assert (sprintf ("%c", repmat ("blah", 2, 1)), "bbllaahh")
 %!assert (sprintf ("%c %c %s", repmat ("blah", 2, 1)), "b b llaahh")
+
+## bug #39735
+%!assert (sprintf ("a %d b", []), "a  b")
+%!assert (sprintf ("a %d b", ''), "a  b")
+%!assert (sprintf ("a %d b", ' '), "a 32 b")
+%!assert (sprintf ("a %s b", []), "a  b")
+%!assert (sprintf ("a %s b", ''), "a  b")
+%!assert (sprintf ("a %s b", ' '), "a   b")
